### PR TITLE
Refactor branch deletion for backports

### PR DESCRIPTION
**Pull request by gaiksaya**
### Description
Refactor branch deletion for backports as current workflow has been failing 2 months. 
https://github.com/opensearch-project/OpenSearch/actions/workflows/delete_backport_branch.yml

The author of the currently used action specified that it is no longer maintained. https://github.com/SvanBoxel/delete-merged-branch/issues/257#issuecomment-2182261868
Hence, refactoring to use GH scripts. Works as expected. See https://github.com/gaiksaya/opensearch-build-libraries/actions/runs/10930031518

### Related Issues
Resolves https://github.com/opensearch-project/OpenSearch/issues/15984

### Check List
- [ ] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: .github/workflows/delete_backport_branch.yml
Comment (peternied) [inline .github/workflows/delete_backport_branch.yml:11]: I'm pretty sure this won't work with the pull request trigger, I think it needs to be updated to pull_request_target.

**Review by peternied — CHANGES_REQUESTED** [review]

File: .github/workflows/delete_backport_branch.yml
Comment (gaiksaya) [reply .github/workflows/delete_backport_branch.yml:11]: Did not change the trigger. It is what previously existed. Pretty sure it works as we want to run this once a pull request in merged (type: closed) in this repository.

**Review by peternied — COMMENTED** [review]
Food for thought; I'd rather we have a sweep and clean job on a cron trigger that looked for merged/closed `backport/` branches and then deleted them that tying this to PR actions.  That would also address all those not-yet-deleted branches that are hanging out.

File: .github/workflows/delete_backport_branch.yml
Comment (peternied) [reply .github/workflows/delete_backport_branch.yml:11]: I'm not so sure, otherwise it would seem that a bad actor change the branch target in a PR and delete main or 2.x on us.  Maybe this works because of the permissions of the bot account or your maintainer account?

Comment (gaiksaya) [reply .github/workflows/delete_backport_branch.yml:11]: The already existing `if` here https://github.com/opensearch-project/OpenSearch/pull/15983/files#diff-0153e43e8683726e844f93b477ceed014452fc9006230d868bb6210bbff54d2aR12 would be the gatekeeper right?

Comment (peternied) [reply .github/workflows/delete_backport_branch.yml:11]: On a PR the author could remove that check, the PR author shouldn't be a gatekeeper for security concerns.  But you know what, either it will work or it won't - lets try this out and if its broken it won't be any worse than before.

**Review by peternied — APPROVED** [review]

**Comment by peternied** [comment]
@gaiksaya I'll merge and lets check back in tomorrow to see if this is working.

**Comment by gaiksaya** [comment]
> @gaiksaya I'll merge and lets check back in tomorrow to see if this is working.

Sure! Thanks! Please merge when you see fit.

**Review by kotwanikunal — APPROVED** [review]

**Comment by gaiksaya** [comment]
Hi @peternied @kotwanikunal Do we need to backport this to 2.x?

**Comment by peternied** [comment]
@gaiksaya Yes it does.  However, I'd rather confirm that it works on backports done from main first.

And unfortunately, it looks like it isn't working, [link](https://github.com/opensearch-project/OpenSearch/actions/runs/10947067185)
image

**Comment by gaiksaya** [comment]
Looks like intendation issue. Fixed here: https://github.com/opensearch-project/OpenSearch/pull/16003
Thanks!

**Comment by gaiksaya** [comment]
Also looks like the workflows in the merged branches are the ones that run. Hence, need to backport in order to delete branches for the PRs that get merged in 2.x

File: .github/workflows/delete_backport_branch.yml
Comment (gaiksaya) [reply .github/workflows/delete_backport_branch.yml:11]: Also noticed one thing. We have branch protection rules in place that do not allow branch deletions by default. Only the branches with `backport/` are allowed to be deleted.

**Comment by peternied** [comment]
@gaiksaya Can you create the backport pr to do this?

**Comment by gaiksaya** [comment]
> @gaiksaya Can you create the backport pr to do this?

Done! https://github.com/opensearch-project/OpenSearch/pull/16008

**Pull request merged**
